### PR TITLE
[SPARK-33052][SQL][TEST] Make all the database versions up-to-date for integration tests

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR intends to update database versions below for integration tests;
 - ibmcom/db2:11.5.0.0a => ibmcom/db2:11.5.4.0 in `DB2[Krb]IntegrationSuite`
 - mysql:5.7.28 => mysql:5.7.31 in `MySQLIntegrationSuite`
 - postgres:12.0 => postgres:13.0 in `Postgres[Krb]IntegrationSuite`
 - mariadb:10.4 => mariadb:10.5 in `MariaDBKrbIntegrationSuite`

Also, this added environmental variables so that we can test with any database version and all the variables are as follows (see documents in the code for how to use all the variables);
 - DB2_DOCKER_IMAGE_NAME
 - MSSQLSERVER_DOCKER_IMAGE_NAME
 - MYSQL_DOCKER_IMAGE_NAME
 - POSTGRES_DOCKER_IMAGE_NAME

### Why are the changes needed?

To improve tests.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually checked.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @HyukjinKwon @dongjoon-hyun

**Review by HyukjinKwon — APPROVED** [review]
I haven't tested by myself but seems fine. BTW, shall we say (at the PR description) the new environment variables exposed in this PR?

**Comment by maropu** [comment]
okay, thanks for the check, @HyukjinKwon

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala [hunk 1]
```diff
@@ -30,7 +30,7 @@ class MariaDBKrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
   override protected val keytabFileName = "mariadb.keytab"
 
   override val db = new DatabaseOnDocker {
-    override val imageName = "mariadb:10.4"
+    override val imageName = sys.env.getOrElse("MARIADB_DOCKER_IMAGE_NAME", "mariadb:10.4")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:33]: Is this the latest one?

**Comment by dongjoon-hyun** [comment]
cc @gaborgsomogyi for Kerberos integration testing.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala [hunk 2]
```diff
@@ -28,7 +28,7 @@ import org.apache.spark.tags.DockerTest
 @DockerTest
 class DB2IntegrationSuite extends DockerJDBCIntegrationSuite {
   override val db = new DatabaseOnDocker {
-    override val imageName = "ibmcom/db2:11.5.0.0a"
+    override val imageName = sys.env.getOrElse("DB2_DOCKER_IMAGE_NAME", "ibmcom/db2:11.5.4.0")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2IntegrationSuite.scala:31]: I verified this manually.
```
DB2IntegrationSuite:
- Basic test
- Numeric types
- Date types
- String types
- Basic write test
- query JDBC option
```

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala [hunk 3]
```diff
@@ -29,7 +29,7 @@ import org.apache.spark.tags.DockerTest
 @DockerTest
 class PostgresIntegrationSuite extends DockerJDBCIntegrationSuite {
   override val db = new DatabaseOnDocker {
-    override val imageName = "postgres:12.0-alpine"
+    override val imageName = sys.env.getOrElse("POSTGRES_DOCKER_IMAGE_NAME", "postgres:13.0-alpine")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresIntegrationSuite.scala:32]: I verified this.
```
PostgresIntegrationSuite:
- Type mapping for various types
- Basic write test
- Creating a table with shorts and floats
- SPARK-20557: column type TIMESTAMP with TIME ZONE and TIME with TIME ZONE should be recognized
- SPARK-22291: Conversion error when transforming array types of uuid, inet and cidr to StingType in PostgreSQL
- query JDBC option
- write byte as smallint
- character type tests
- SPARK-32576: character array type tests
```

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresKrbIntegrationSuite.scala [hunk 4]
```diff
@@ -30,7 +30,7 @@ class PostgresKrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
   override protected val keytabFileName = "postgres.keytab"
 
   override val db = new DatabaseOnDocker {
-    override val imageName = "postgres:12.0"
+    override val imageName = sys.env.getOrElse("POSTGRES_DOCKER_IMAGE_NAME", "postgres:13.0")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/PostgresKrbIntegrationSuite.scala:33]: I verified this, too.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala [hunk 5]
```diff
@@ -27,7 +27,7 @@ import org.apache.spark.tags.DockerTest
 @DockerTest
 class MySQLIntegrationSuite extends DockerJDBCIntegrationSuite {
   override val db = new DatabaseOnDocker {
-    override val imageName = "mysql:5.7.28"
+    override val imageName = sys.env.getOrElse("MYSQL_DOCKER_IMAGE_NAME", "mysql:5.7.31")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MySQLIntegrationSuite.scala:30]: This is verified.
```
MySQLIntegrationSuite:
- Basic test
- Numeric types
- Date types
- String types
- Basic write test
- query JDBC option
```

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala [hunk 6]
```diff
@@ -27,7 +27,8 @@ import org.apache.spark.tags.DockerTest
 @DockerTest
 class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
   override val db = new DatabaseOnDocker {
-    override val imageName = "mcr.microsoft.com/mssql/server:2019-GA-ubuntu-16.04"
+    override val imageName = sys.env.getOrElse("MSSQLSERVER_DOCKER_IMAGE_NAME",
+      "mcr.microsoft.com/mssql/server:2019-GA-ubuntu-16.04")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:31]: This is not changed.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2KrbIntegrationSuite.scala [hunk 7]
```diff
@@ -35,7 +35,7 @@ class DB2KrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
   override protected val keytabFileName = "db2.keytab"
 
   override val db = new DatabaseOnDocker {
-    override val imageName = "ibmcom/db2:11.5.0.0a"
+    override val imageName = sys.env.getOrElse("DB2_DOCKER_IMAGE_NAME", "ibmcom/db2:11.5.4.0")
```
Comment (dongjoon-hyun) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/DB2KrbIntegrationSuite.scala:38]: This is verified.
```
DB2KrbIntegrationSuite:
Java config name: /Users/dongjoon/PRS/SPARK-REVIEW/external/docker-integration-tests/target/tmp/spark-37736442-df33-4ba9-a4e5-edb2df23aee9/1601622016820/krb5.conf
Loaded from Java config
```

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:33]: This is not changed, but I failed. So, cc @gaborgsomogyi 
```
MariaDBKrbIntegrationSuite:
org.apache.spark.sql.jdbc.MariaDBKrbIntegrationSuite *** ABORTED ***
  Exception encountered when invoking run on a nested suite - The code passed to eventually never returned normally. Attempted 26 times over 2.269450662233333 minutes. Last failure message: Could not connect to address=(host=192.168.1.114)(port=63328)(type=master) : Socket fail to connect to host:192.168.1.114, port:63328. connect timed out. (DockerJDBCIntegrationSuite.scala:159)
```

Comment (maropu) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:33]: This test with the latest one, 10.5, failed consistently, but 10.4 looks fine on my local env.

**Comment by dongjoon-hyun** [comment]
Thank you for this updates, @maropu . I verified mostly. For the rest, I'll leave it to you and @gaborgsomogyi .

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala (hunk 1 above)
Comment (maropu) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:33]: Could you check this, @gaborgsomogyi ?

**Comment by maropu** [comment]
Thanks for your manual checks, @dongjoon-hyun !

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala (hunk 1 above)
Comment (gaborgsomogyi) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:33]: Started to have a look...

Comment (gaborgsomogyi) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:40]: I've executed the mentioned test with this PR but it works fine on my machine.
Screenshot 2020-10-02 at 10 23 09
What I can imagine is that the docker version is different. When I've started the development in this area I remember there was quite a stuggle to install a working version. I'm still not on the latest.
Screenshot 2020-10-02 at 10 17 49

Comment (maropu) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:40]: Ah, ok. Thanks for the check. I will re-run it on the latest one.

Comment (gaborgsomogyi) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:40]: I've had a look at why `10.5` is failing. Please see that the gssapi plugin version is pinned to a specific version: https://github.com/apache/spark/blob/9b88aca2954cd931c94a7cc788c3c3f7a33e99b7/external/docker-integration-tests/src/test/resources/mariadb_docker_entrypoint.sh#L21

When the version changed to `1:10.5.5+maria~focal` then it works like charm :)

**Comment by gaborgsomogyi** [comment]
Double checked and LGTM, except I suggest to bump the MariaDB version to 10.5 w/ the suggested change.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala (hunk 1 above)
Comment (maropu) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:40]: Oh, I see! Thanks a lot! I'll check and update it.

**Comment by maropu** [comment]
okay, I've checked the latest mariadb in `MariaDBKrbIntegrationSuite` worked well on my local env. Anyone can check this again? Thanks! @HyukjinKwon @dongjoon-hyun @gaborgsomogyi

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala [hunk 8]
```diff
@@ -30,7 +30,9 @@ class MariaDBKrbIntegrationSuite extends DockerKrbJDBCIntegrationSuite {
   override protected val keytabFileName = "mariadb.keytab"
 
   override val db = new DatabaseOnDocker {
-    override val imageName = "mariadb:10.4"
+    // If you change `imageName`, you need to update the version of `mariadb-plugin-gssapi-server`
+    // in `resources/mariadb_docker_entrypoint.sh` accordingly.
```
Comment (gaborgsomogyi) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:34]: I think it's doable to automate it with an additional env var but it's a nice to have in my view.

**Review by gaborgsomogyi — APPROVED** [review]
LGTM, all tests passed.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala (hunk 8 above)
Comment (maropu) [reply external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MariaDBKrbIntegrationSuite.scala:34]: Yea, it looks nice if we can automate it.

**Review by srowen — APPROVED** [review]

**Comment by maropu** [comment]
Thanks, @srowen ! Merged to master.

**Comment by dongjoon-hyun** [comment]
Thank you, @maropu and all!

**Comment by dongjoon-hyun** [comment]
Hi, @maropu .
Could you resolve the JIRA issue, SPARK-33052?

**Comment by maropu** [comment]
Oh, I forgot to do so. Thanks for letting me know, @dongjoon-hyun

**Review by HyukjinKwon — COMMENTED** [review]
+1 lgtm

**Pull request closed without merging**